pkg/kubelet/logs/container_log_manager.go

```diff
@@ -176,19 +176,107 @@ func (c *containerLogManager) Start() {
 	}, logMonitorPeriod)
 }
 
+type containerWithLog struct {
+	ID        string
+	CreatedAt int64
+	logPath   string
+	State     runtimeapi.ContainerState
+}
+
+func (c *containerLogManager) addContainerToMap(container *runtimeapi.Container, path string, containersByName map[string][]containerWithLog) {
+	if container.Metadata != nil {
+		index := -1
+		for i := range containersByName[container.Metadata.Name] {
+			if containersByName[container.Metadata.Name][i].ID == container.Id {
+				index = i
+				containersByName[container.Metadata.Name][i].ID = container.Id
+				containersByName[container.Metadata.Name][i].State = container.State
+				containersByName[container.Metadata.Name][i].CreatedAt = container.CreatedAt
+				break
+			}
+		}
+		if index < 0 {
+			withLog := containerWithLog{
+				ID:        container.Id,
+				State:     container.State,
+				CreatedAt: container.CreatedAt,
+				logPath:   path,
+			}
+			containersByName[container.Metadata.Name] = append(containersByName[container.Metadata.Name], withLog)
+		} else if containersByName[container.Metadata.Name][index].logPath == "" {
+			containersByName[container.Metadata.Name][index].logPath = path
+		}
+	}
+}
+
+func (c *containerLogManager) pruneDeadContainerLogs(containersByName map[string][]containerWithLog) {
+	containersToKeep := 2
+	for k, containers := range containersByName {
+		sort.Sort(containersByCreation(containers))
+		var logPath string
```

Comment:
Is it specified in CRI documentation that logs of all instances of one container kept under the same path?

What forbids containerd or something else keep logs as:
/var/log/pods/<podUID>/containerName/{id}/container.log instead of
/var/log/pods/<podUID>/containerName/{id}.log

Reply:
Log rotation is done by container_log_manager.
The logs (original and rotated) are kept in the same directory.

Reply:
For single run of container yes. But if after each restart logs are going to be written to different directory, then logs of previous containers are not going to be removed, because they are stored in different place.

Reply:
Can you give me an example output where the log of restarted container is stored in different directory ?

Reply:
I'll need to implement my own container runtime. I'm just saying that you rely on fact, that is not documented in CRI. Or at least I haven't managed to find it.

Reply:
I don't think every piece of code for container handling in kubelet has CRI documentation written behind it.
When you find such container runtime, please let me know.

Reply:
Is docker good enough? 
Logs are written to: /var/lib/docker/containers/{container-id}/{container-id}-json.log file. And even worse, that directory contains bunch of other files:

`sudo ls /var/lib/docker/containers/740ad7fc986c9791f8c46f0eaae921b820aa1bbb5a20212c48fcca225b5e0b6b`

> 740ad7fc986c9791f8c46f0eaae921b820aa1bbb5a20212c48fcca225b5e0b6b-json.log    checkpoints      mounts
> 740ad7fc986c9791f8c46f0eaae921b820aa1bbb5a20212c48fcca225b5e0b6b-json.log.1  config.v2.json
> 740ad7fc986c9791f8c46f0eaae921b820aa1bbb5a20212c48fcca225b5e0b6b-json.log.2  hostconfig.json
>

Reply:
And if you know about places where kubelet relies on undocumented behavior of CRI, please file an issue.

Reply:
The logs Matt and I looked at are located under /var/log/pods/.

Can you reveal the contents of hostconfig.json and config.v2.json ?

Reply:
There is no guarantee that logs are going to be stored under /var/log/pods/. Is content of those files important? Below I'm providing content of hostconfig.json. But my point that this PR is built on too many assumptions about container runtimes, which is not correct way of solving this problem, otherwise we are going to get dozen new issues if any new runtime will just follow only documented behavior.

Hostconfig example:

> {"Binds":["/etc/ssl/certs:/etc/ssl/certs","/var/lib/kubelet/pods/3a7c248d-a03d-449a-a0f6-55cd3cc67b50/volumes/kubernetes.io~secret/metadata-agent-token-22zsv:/var/run/secrets/kubernetes.io/serviceaccount:ro","/var/lib/kubelet/pods/3a7c248d-a03d-449a-a0f6-55cd3cc67b50/etc-hosts:/etc/hosts","/var/lib/kubelet/pods/3a7c248d-a03d-449a-a0f6-55cd3cc67b50/containers/metadata-agent/04356ee4:/dev/termination-log"],"ContainerIDFile":"","LogConfig":{"Type":"json-file","Config":{"max-file":"5","max-size":"10m"}},"NetworkMode":"container:439f391fe2c86578264fbd98ae460ed4e8e249617e7ec4a611c6fc660904fbe7","PortBindings":null,"RestartPolicy":{"Name":"no","MaximumRetryCount":0},"AutoRemove":false,"VolumeDriver":"","VolumesFrom":null,"CapAdd":null,"CapDrop":null,"Capabilities":null,"Dns":null,"DnsOptions":null,"DnsSearch":null,"ExtraHosts":null,"GroupAdd":null,"IpcMode":"container:439f391fe2c86578264fbd98ae460ed4e8e249617e7ec4a611c6fc660904fbe7","Cgroup":"","Links":null,"OomScoreAdj":-998,"PidMode":"","Privileged":false,"PublishAllPorts":false,"ReadonlyRootfs":false,"SecurityOpt":["seccomp=unconfined"],"UTSMode":"","UsernsMode":"","ShmSize":67108864,"Runtime":"runc","ConsoleSize":[0,0],"Isolation":"","CpuShares":49,"Memory":100663296,"NanoCpus":0,"CgroupParent":"/kubepods/pod3a7c248d-a03d-449a-a0f6-55cd3cc67b50","BlkioWeight":0,"BlkioWeightDevice":null,"BlkioDeviceReadBps":null,"BlkioDeviceWriteBps":null,"BlkioDeviceReadIOps":null,"BlkioDeviceWriteIOps":null,"CpuPeriod":100000,"CpuQuota":4800,"CpuRealtimePeriod":0,"CpuRealtimeRuntime":0,"CpusetCpus":"","CpusetMems":"","Devices":[],"DeviceCgroupRules":null,"DeviceRequests":null,"KernelMemory":0,"KernelMemoryTCP":0,"MemoryReservation":0,"MemorySwap":100663296,"MemorySwappiness":null,"OomKillDisable":false,"PidsLimit":null,"Ulimits":null,"CpuCount":0,"CpuPercent":0,"IOMaximumIOps":0,"IOMaximumBandwidth":0,"MaskedPaths":["/proc/acpi","/proc/kcore","/proc/keys","/proc/latency_stats","/proc/timer_list","/proc/timer_stats","/proc/sched_debug","/proc/scsi","/sys/firmware"],"ReadonlyPaths":["/proc/asound","/proc/bus","/proc/fs","/proc/irq","/proc/sys","/proc/sysrq-trigger"]}

Reply:
Not sure if you saw my modified comment above.
Since we are handling not just the most recently active log file, there has to be an agreement on how the filtering is done.
Current proposal is to filter out filename without ".log" in it.

Reply:
No I have seen previous version of your comment.

That will fail for files that are named .login, .logo, and few hundreds other. And it's not going to work when file doesn't have .log. Nothing stops docker from changing file name from {id}-json.log to {id}-log.json. 

And we still have an open problem, that files are not located in the same directory. So there is a chance that container_log_manager is not going to get path from which it should remove files at all.

Reply:
I didn't see files named .login, .logo in your above posting.
If these files were in the directory, please post another listing.
```
from {id}-json.log to {id}-log.json
```
json file is normally for configuration.
I wonder why docker would make such a change.

Reply:
Interesting that resolv.conf shows up along with the log file in the same docker directory (from sample listing I obtained).
But I can tell that there is no .login or .logo file in the same dir.

Reply:
It's just an example. 

1. You are assuming that no other files are going to be created by container runtime with .log substring.
2. You are assuming that log files are going to have .log extension, which is not specified by the API. It can easily be -log or etc.
3. You are still assuming that all logs are stored in the same directory. I have shown on Docker example that this is incorrect.